### DISCUSSION TITLE
[Testing] NineChronicles.Headless 실행 옵션 중 --store-type 을 String > Enum type 으로 변경하자

### DISCUSSION THREAD

**Discussion by longfin** [planetarium/NineChronicles.Standalone#1449] [Ideas]
https://github.com/planetarium/NineChronicles.Headless/issues/1438 를 작업하다 든 생각입니다. — @U-lis 

1. `NineChronicles.Headless`에서는 `StoreType.cs`에서 store type 을 Enum 으로 정의하고, `Tip`, `Inspect` command에서 사용하고 있습니다
2. `Libplanet.Headless`에서는 store type 을 사용자 입력값 그대로 문자열로 받아 이에 따라 `LoadStore`로직을 구분합니다. ([https://github.com/planetarium/NineChronicles.Headless/blob/e97ec44689dab8fb895ee0d3e99704ac9ff67936/Libplanet.Headless/Hosting/LibplanetNodeService.cs#L291-L294](https://github.com/planetarium/NineChronicles.Headless/blob/e97ec44689dab8fb895ee0d3e99704ac9ff67936/Libplanet.Headless/Hosting/LibplanetNodeService.cs#L291-L294))
3. 저는 StoreType 같은 것은 선택지가 명확하기 때문에 Enum 으로 모두 관리하는 것이 맞지 않나 생각해 제안을 드렸습니다.

---

- @Akamig , @BasixKOR님의 의견은 Enum 으로 하게 되면 다른 사람이 자기가 원하는 종류의 store type 을 이거저거 시도하려면 Enum 부터 손봐야 해서 새 store type 추가에 대해 자유도를 주기 위해 일부러 `String`을 사용하는 것일 수 있다는 의견을 주셨습니다.
    - @BasixKOR (의견 보충) Libplanet 단에서 String으로 처리하는 이유에 대한 설명이었는데, NineChronicles.Headless의 경우 개발자보다는 end-user가 실행하는 걸 염두에 둔 제품인 만큼 enum으로 두는 것도 바람직해 보입니다. 즉 Libplanet 내에서는 새로운 IStore 확장을 고려하여 확장 가능한 형태로 유지하되 Headless 내에서는 편의에 맞춰 Enum 등으로 열거하는 정도면은 적절한 선일 것 같아요.
    - @Akamig  Enum도 새로운 store type을 도입하는 경우에 그렇게 확장성이 나쁘지 않…나…? 그냥 예를 들어 뭐 RocksDB 대신 CockroachDB를 도입한다고 했을때 개발하는 입장에서 자유도가 그렇게 떨어지냐 하면 그건 또 아닌것 같고…?

다른 분들의 의견이나, 이 부분에 대해 제가 고려하지 못하고 있는 history 등이 있다면 말씀 부탁드립니다.

- @moreal 헤드리스 레벨에서는 사용하는 Store의 타입이 정해져 있고 StoreType enum으로 만들어 제공해주는 게 맞는 것 같습니다.
- @dahlia 최근에 Libplanet 쪽에서 이런 기능이 추가되었는데 살펴보셔도 좋을 것 같습니다. https://github.com/planetarium/libplanet/pull/1955
    - 제 개인적인 의견으로는, 장기적으로는 저장소의 백엔드 종류와 저장소의 경로 등에 대한 설정을 모두 URI 형태로 표현 가능하게 하는 게 좋을 것 같습니다. 왜냐하면 저장소 백엔드 중에서는 “경로” 같은 게 의미 없는 경우도 있을 수 있고(이를테면 인메모리 저장소 백엔드나, RDBMS 저장소 등이 이론상 존재 가능) 경로와는 또다른 부가적인 옵션이 요구될 수도 있습니다.
    그럴 때마다 “`--store-path` 옵션은 `--store-type`이 `rocksdb`이거나 `file`일 때만 쓸 수 있음”, “`--store-host` 및 `--store-port` 옵션은 `--store-type`이 `mysql`일 때만 쓸 수 있음”과 같은 단서를 추가하면서 `--store-*` 옵션을 늘리는 것보다는 `--store-uri` 같은 하나의 옵션으로 받을 수 있게 하는 쪽이 더 좋지 않을까요?
        - @BasixKOR 이 방향으로 개발이 된다면 여러 Store 구현체가 사용하는 리소스가 동일한 경우 scheme으로 구분되어야 할 것 같네요. 예를 들어 file:///var/run/nine-chronicles/9c-main 같은 경우에는 LiteDB로 해석해야 할 지 아니면 RocksDB로 해석해야 할 지 등이 갈리는 문제가 있을 듯 합니다. — rocksdb+file 같은 식으로 대응되어 있군요!
            - @BasixKOR 근데 요게 MemoryStore에서는 Path가 어떻게 들어가야 하나요?
                - → @U-lis  `memory:///` 로 쓰는 것을 자주 보았습니다.
        - @moreal 이 방향이 참 좋은 것 같아요. 기존 것을 기존 사용자들과 잘 이야기하면서 deprecation 하는 게 문제일 것 같아요.
- @longfin  Interface level 로 먼저 조사해서 사용처를 확인하고, 그 뒤에 이걸 반영해서 실제 option 을 손볼 수 있도록 하면 좋겠습니다.

**Comment by moreal** [comment]
Related issues for deprecation: #1447
